File: Template/DataStructure/BitwiseTrie.py
"""
Bitwise Trie

Example:
- CSES-1655 Maximum Xor Subarray
"""

class TrieNode:
    __slots__ = ['ch']
    def __init__(self):
        self.ch = [None, None]

class BitwiseTrie:
    def __init__(self, bitlen=30):
        self.root = TrieNode()
        self.B = bitlen

    def insert(self, x: int):
        node = self.root
        for k in range(self.B, -1, -1):
            b = (x >> k) & 1
            if node.ch[b] is None:
                node.ch[b] = TrieNode()
            node = node.ch[b]

    def max_xor(self, x: int) -> int:
        node = self.root
        res = 0
        for k in range(self.B, -1, -1):
            b = (x >> k) & 1
            if node.ch[b ^ 1] is not None:
                res |= (1 << k)
                node = node.ch[b ^ 1]
            else:
                node = node.ch[b]
        return res

File: Template/DataStructure/BitwiseTrie.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Bitwise Trie

Example:
- CSES-1655 Maximum Xor Subarray
*/

class TrieNode {
public:
    TrieNode* ch[2];
    TrieNode() {
        ch[0] = ch[1] = nullptr;
    }
};

class BitwiseTrie {
public:
    TrieNode* root;
    int B;
    BitwiseTrie(int bitlen = 30) {
        root = new TrieNode();
        B = bitlen;
    }

    void insert(int x) {
        TrieNode* node = root;
        for (int i = B; i >= 0; i--) {
            int b = (x >> i) & 1;
            if (node->ch[b] == nullptr)
                node->ch[b] = new TrieNode();
            node = node->ch[b];
        }
    }

    int max_xor(int x) {
        TrieNode* node = root;
        int res = 0;
        for (int i = B; i >= 0; i--) {
            int b = (x >> i) & 1;
            if (node->ch[b ^ 1] != nullptr) {
                res |= (1 << i);
                node = node->ch[b ^ 1];
            } else {
                node = node->ch[b];
            }
        }
        return res;
    }
        
};